internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,185 @@
+package ratelimit
+
+import (
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Ratelimiter holds all ratelimit buckets
+type RateLimiter struct {
+	sync.Mutex
+	buckets         map[string]*Bucket
```

Comment:
Might be better to have:
global *Bucket

instead of passing the rate limiter into a bucket, circular references feel wrong.